daemon/watchdog.c

```diff
@@ -320,22 +320,20 @@ do_hypercall(privcmd_hypercall_t *hypercall, void *arg, size_t argsize, int *ioc
     else
     {
         ret = ioctl(hypercall_fd, IOCTL_PRIVCMD_HYPERCALL, hypercall);
-    }
-    if (ret < 0)  
-    {
-
-        if (currentstatus == MTC_SUCCESS)
+        if (ret < 0 && currentstatus == MTC_SUCCESS)
         {
-            log_internal(MTC_LOG_ERR, "WD: hypercall failed (sys %d).\n", ret);
+            log_internal(MTC_LOG_ERR, "WD: hypercall failed (sys %d).\n", errno);
         }
-
-        unlock_pages(hypercall, sizeof(privcmd_hypercall_t));
-        unlock_pages(arg, argsize);
-        return MTC_ERROR_WD_INSTANCE_UNAVAILABLE ;
     }
+
     unlock_pages(hypercall, sizeof(privcmd_hypercall_t));
     unlock_pages(arg, argsize);
 
+    if (ret < 0)
```

Comment:
Did you intend to move these around? I'd keep changes minimal.

Reply:
In all 3 cases (wd.instance.unavailable; hypercall fails; hypercall returns the watchdog id) the pages need to be unlocked, in the 2 cases where ret is an error the return is the same.

In practice the logging line was moved inside the ioctl block, with a correctly set condition; and the repeated `unlock_pages` were removed.

Without the second change that piece of code would look like this:
```
    if (ret < 0)  
    {
        unlock_pages(hypercall, sizeof(privcmd_hypercall_t));
        unlock_pages(arg, argsize);
        return MTC_ERROR_WD_INSTANCE_UNAVAILABLE ;
    }
    unlock_pages(hypercall, sizeof(privcmd_hypercall_t));
    unlock_pages(arg, argsize);
```
I can leave it like that if the code changes are to be really minimal